### PR TITLE
Sidekiq directly

### PR DESCRIPTION
**Pull request by andycroll**

### CODE REVIEW

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+category: ruby
+image:
+  base: '2021/use-sidekiq-directly-not-through-active-job'
+  alt: "Silhouetted karate kick on the beach"
+  credit: "Jason Briscoe"
+  source: "https://unsplash.com/photos/HN_4K2diUWs"
+
+---
+
+If you build a web application you should minimize the amount of time spent responding to every user; a fast website means a happy user. One way of doing that is to run long-running, parallelisable, or potentially slow, work asynchronously outside of the immediate web request. This might be sending emails, scheduled cleanups, long running calculations or anything that uses an external API.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: Oxford comma after 'calculations', please!

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+[Active Job](https://guides.rubyonrails.org/active_job_basics.html) is the recommended way of enqueuing and running background work in Rails. It provides a wrapper around a bunch of [different implementations](https://api.rubyonrails.org/classes/ActiveJob/QueueAdapters.html) of the concept. Each implementation utilises a different underlying technology and has it's own pros and cons. Most of the implementations existed before [Active Job was released](https://guides.rubyonrails.org/4_2_release_notes.html#active-job) in Rails 4.2.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *its own

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+One of the most popular, battle-tested and performant asynchronous frameworks is [Sidekiq](https://sidekiq.org), which is multi-threaded and utilises Redis for it's queuing storage. It’s an open source library with two paid tiers for additional high-end features.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: Oxford comma in the first list, please!

Comment (nodunayo) [reply _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *its

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+Disclaimer: I'm a massive fan and “Pro” customer of Sidekiq and have met [Mike](https://twitter.com/getajobmike) (it's author) a little at a couple of Ruby conferences. That hasn't coloured my view of the library, it has _always_ been terrific and is worth every penny for our use at [work](https://coveragebook.com).
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *_its_ author...

*remove 'a little' before 'at a couple...'

*Full stop after 'view of the library.

*worth every penny given how we use it at work

(Maybe this disclaimer should go at the end??)

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    an_active_record_object = ActiveRecordObject.find_by(id: id)
+    an_active_record_object.do_things
+  end
+end
+```
+
+
+## Why?
+
+Active Job is a solid super-set of Sidekiq's functionality, but you’re missing out on a number of advantages if you use it without.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *superset ?

*if you don't use Sidekiq directly

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+If you're doing _a lot_ of work in the background, enqueuing many small fast-running jobs (as is generally suggested) there is a [2—20x performance improvement](https://github.com/mperham/sidekiq/wiki/Active-Job#performance) if you use Sidekiq directly. This improvement will be dependant on your specific setup.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *comma needed after closing parenthesis 

*The level of improvement will be dependant...

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+If you are enqueing lots of jobs, you could be using [Sidekiq’s bulk enqueing functionality](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing), this is hard to do performantly when using sidekiq with ActiveJob.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: Full stop after closing parentheses.

Comment (nodunayo) [reply _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: Capitalise Sidekiq.

And decide if you're saying 'Active Job' with a space or not.

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+When using Sidekiq via Active Job there are, confusingly, multiple levels of retries in the event of a job failure. Active Job has it's own retry mechanism, which once completed then passes down to Sidekiq’s own (seperate and different!) retry system. This is a strange thing to debug when things go wrong.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *its

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+When using Active Job and passing in an Active Record object, you rely on the object being serialized to a text argument using [Global ID](https://github.com/rails/globalid). This does save doing the lookup yourself (as you can see in the Active Job example) but it raises errors if records can be deleted (or even not created!) before the job is pulled form the queue. It also makes the arguments of the jobs a bit ugly in the web dashboard.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *from

This paragraph doesn't flow well.

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+There’s no real harm to “locking yourself into” a specific job framework. You are unlikely to swap in future for another queuing system with similar (intentionally) limited functionality availible in Active Job. In a real life situation where your asyncronous infrastructure is struggling, you’d be assessing various solutions with more functionality and planning for a major migration project.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: I think quotes should only be around "locking"

*You are unlikely to switch to another queuing system...

End of this paragraph feels messy/out of place.

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+If you're using Sidekiq in prduction you should buy a license for the Pro version. The [extra reliability](https://github.com/mperham/sidekiq/wiki/Reliability#using-super_fetch) so that jobs are not “lost” due to crashes is enough of a reason when in a production environment, but there are other additional features including job deletion (by job id and kind of job) and web dashboard enhancements.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *production

Seems like there are extra spaces in between 'so that jobs are not lost'

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+
+## Why not?
+
+When using Sidekiq directly you have to think more about your job arguments and perform object lookups yourself, Sidekiq only permits simple values as job arguments.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: Full stop before Sidekiq.

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+You can use Active Job-style _and_ direct Sidekiq jobs onto the same queues at the same time because Active Job wraps Sidekiq. You don't have to switch everything, you can use both styles in the same application.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: Messy paragraph.

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+There is the deeper question as to why use Sidekiq at all if your workload is light, or you are early on in your project.
+
+[Good Job](https://github.com/bensheldon/good_job) and [Que](https://github.com/que-rb/que) (postgres only) or [Delayed Job](https://github.com/collectiveidea/delayed_job) (any SQL database) are well-regarded Active Job adapters and all use your existing database (in better or worse ways depending or your specific application) and then you don’t need to run an additional piece of database infrastructure.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: After the parentheses, start a new sentence: If you use one of these, you then don't need to run...

**Review by nodunayo — DISMISSED** [review]
This piece feels like it's lacking in flow. Feels like a bunch of random points put one after the other, and the paragraphs aren't always clear/easy to read.

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+    an_active_record_object = ActiveRecordObject.find_by(id: id)
+    an_active_record_object.do_things
+  end
+end
+```
+
+
+## Why?
+
+You’re missing out on a number of advantages that you gain when you use Sidekiq directly.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]:
```suggestion
You’re missing out on a number of advantages when you don't use Sidekiq directly.
```

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+If you're doing _a lot_ of work in the background, enqueuing many small fast-running jobs—as is generally suggested—there is a [2—20x performance improvement](https://github.com/mperham/sidekiq/wiki/Active-Job#performance) if you use Sidekiq directly. The level of improvement will be dependant on your specific setup.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *leads to a 2-20x perf..

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+If you are enqueing lots of jobs, you could be using [Sidekiq’s bulk enqueing functionality](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing), this is hard to do performantly when using Sidekiq with Active Job.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]:
```suggestion
If you are enqueing lots of jobs, you could use [Sidekiq’s bulk enqueing functionality](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing). This is hard to do performantly when using Sidekiq with Active Job.
```

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+There are, confusingly, multiple levels of retries in the event of a job failure. Active Job has its own retry mechanism, which once completed then passes down to Sidekiq’s own (seperate and different!) retry system. This is a confusing thing to debug when things go wrong.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: Maybe take out the first 'confusingly' or change the last one to "tricky" or "difficult" or something like that.

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+Active Job allows passing in an Active Record object to the `#perform` method, which is then serialized to a text argument using [Global ID](https://github.com/rails/globalid). This saves doing the lookup yourself but it can raise errors if records are deleted before the job is pulled form the queue. This automatic serialization also makes the arguments to the Sidekiq jobs difficult to read in the web dashboard.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: *from

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+
+There’s no real harm to “locking yourself into” a specific job framework. You are unlikely to swap to another queuing system with similarly limited Active Job functionality. When you consider replacing your asyncronous infrastructure you should assess various solutions and expect a major migration project.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: This paragraph could be a little clearer. The first part, I think you need to make it clearer that you're trying to say: if using Sidekiq directly, don't worry about locking yourself into something specific.

The second part: are you saying that if you're replacing your *_asynchronous_ (sp) infrastructure, then you should be considering a migration then regardless?

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+
+## Why not?
+
+When using Sidekiq directly you have to think more about your job arguments and perform object lookups yourself. Sidekiq only permits simple values as job arguments.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]:
```suggestion
When using Sidekiq directly you have to think more about what your job arguments should be and then perform object lookups yourself. Sidekiq only permits simple values as job arguments.
```

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+
+You can define and use Active Job-style _and_ direct Sidekiq jobs at the same time. You don't have to switch, you can use both styles in the same application.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: Why is this part of "Why not"?

*You don't have to use one or the other. Both styles can be used in the same application.

(Is this recommended?)

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+
+There is a deeper question as to why use Sidekiq at all if your workload is light, or you are early on in your project.
```
Comment (nodunayo) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]: Maybe rephrase this to say something like: if your workload is like or you're early on in your project, perhaps you can start with AJ/you don't need to worry about setting up Sidekiq.

**Review by nodunayo — DISMISSED** [review]

File: _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+If you _are_ replacing your asynchronous infrastructure you should expect a major migration project, rather than just switching out an adapter via a single configuration line.
+
+
+## Why not?
+
+When using Sidekiq directly you have to think more about what your job arguments should be and then perform object lookups yourself. Sidekiq only permits simple values as job arguments.
+
+There’s no need to change all your existing jobs if you are already using Sidekiq through Active Job. You don't have to use one or the other, you can define and use both at the same time. This might be a little confusing to manage, but it isn’t a terrible solution. You can just use direct Sidekiq jobs when you need the improved performance and bulk-enqueing.
+
+If your workload is light, or you are early on in your project, you might not need Sidekiq (at least its Redis dependancy) at all.
```
Comment (andycroll) [inline _posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md]:
```suggestion
If your workload is light, or you are early on in your project, you might not need Sidekiq (or at least its Redis dependancy) at all.
```
